src/domain/consents/{ID}/generateChallenge.ts

```diff
@@ -0,0 +1,111 @@
+ Copyright © 2020 Mojaloop Foundation
+ The Mojaloop files are made available by the Mojaloop Foundation under the
+ Apache License, Version 2.0 (the 'License') and you may not use these files
+ except in compliance with the License. You may obtain a copy of the License at
+ http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, the Mojaloop files
+ are distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied. See the License for the specific language
+ governing permissions and limitations under the License.
+ Contributors
+ --------------
+ This is the official list of the Mojaloop project contributors for this file.
+ Names of the original copyright holders (individuals or organizations)
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ - Abhimanyu Kapur <abhi.kapur09@gmail.com>
+
+ --------------
+ ******/
+
+import { Request } from '@hapi/hapi'
+import { consentDB } from '../../../lib/db'
+import { thirdPartyRequest } from '../../../lib/requests'
+import { Consent } from '../../../model/consent'
+import { Enum } from '@mojaloop/central-services-shared'
+import SDKStandardComponents from '@mojaloop/sdk-standard-components'
+import { ExternalScope } from '../../../lib/scopes'
+
+/*
+ * Interface for Consent Credential resource type
+ */
+export interface ConsentCredential {
+  credentialType: string;
+  credentialStatus: string;
+  credentialPayload?: string | undefined;
+  credentialChallenge: string;
+}
+
+/**
+ * Validates whether generate challenge request is valid
+ * by comparing consent ID sent matches with existing consent in table
+ * and if source ID matches initiator ID
+ * @param request: request received from PISP
+ * @param consent: Consent object
+ */
+export function isConsentRequestInitiatedByValidSource (
+  request: Request,
+  consent: Consent): boolean {
+  const fspiopSource = request.headers[Enum.Http.Headers.FSPIOP.SOURCE]
+  return (consent && consent.initiatorId === fspiopSource)
+}
+
+/**
+ * Assigns credentials to given consent object and updates in the database
+ * Returns updated consent object
+ */
+export async function updateConsentCredential (
+  consent: Consent,
+  credential: ConsentCredential): Promise<Consent> {
+  // Update consent credentials
+  consent.credentialType = credential.credentialType
+  consent.credentialStatus = credential.credentialStatus
+  consent.credentialChallenge = credential.credentialChallenge
+
+  // Update in database
+  await consentDB.update(consent)
```

Comment:
does it update only `credential*` related fields or all consent fields?

Reply:
maybe we should implement a guard here to be sure which fields of consent are updated

Reply:
I believe #21 by @ramanmangla  deals with this